ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -105,10 +105,14 @@ data NoSuchHonestChainSchema =
 genHonestRecipe :: QC.Gen HonestRecipe
 genHonestRecipe = sized1 $ \sz -> do
     (Kcp k, Scg s, Delta d) <- genKSD
-    -- 2s slots has at least 2k blocks. But that is ensuring k-1 more blocks
+    -- 2s slots have at least 2k blocks. But that is ensuring k-1 more blocks
     -- than we actually need. Thus it's safe to remove the k-1 last slots.
-    -- Therefore we set for a length of at least 2s - (k - 1).
-    l <- (+ (2*s - (k - 1))) <$> QC.choose (0, 5 * sz)
+    -- Therefore we first set for a length of at least 2s - (k - 1).
+    --
+    -- If we then add k+d slots (to get a total size of 2s+d+1), we get enough
```

Comment:
I'm not remembering why adding `k+d` slots ensures room for `k+1` adversarial active slots after the intersection.

What I do recall is that the basic idea is to ensure there are enough slots after the stability window anchored in the first adversarial active slot. I _think_ "enough" means `k`, since the first adversarial active slot that anchors the delimiting window counts as the `+1`.

The following is I've come up with (after writing PR https://github.com/IntersectMBO/ouroboros-consensus/pull/897 in preparation)---though I'm not constraining my reasoning here with the assumption that `uniformAdversarialChain` can't change.

I'm hoping I'm missing something, and the reasoning you've already done is simpler :/ At least my counts are similar to the one you have.

----

Let sai abbreviate "slot after the intersection".
Let asai abbreviate "active sai".

`uniformAdversarialChain` requires that the honest schema permits an intersection with both `Kcp+1` honest asais and also enough sais to also allow at least `Kcp+1` alternative asais.
Specifically, there must be enough sais so that at least `Kcp` slots are _unstable_, ie both more than `Del` after the `Kcp+1`st honest asai and also more than `Scg` after the first alternative asai.

Supose `1<Kcp`.
The `Kcp+1`st honest asai can be no later than the `2Scg-(Kcp-1)`th sai.
The unstable slots could begin as late as `Del+1` after that.
The first alternative asai is unconstrained when `1<Kcp`, and thus merely needs to be no later than the `2Scg-(Kcp-1)+Del+1`st sai to ensure there are at least `Kcp` unstable slots before the end of the schema.
Thus `2Scg+Del+2` sais would suffice to ensure there are at least `Kcp` unstable slots before the end of the schema.

Suppose `Kcp=0`.
Note that there does not need to be an unstable slots, merely at least one alternative asai.
The `Kcp+1`st honest asai is the first honest sai, and so can be no later than the `Scg`th sai, since Genesis requires the first `Scg` sais contains _strictly_ more honest active than alternative active.
(This requires that the honest schema has two active slots in the same window of `Scg+1` slots, one of which may be the genesis block.)
Thus it's possible that the first alternative asai can be no sooner than `Scg+Del+1`.
Thus `Scg+Del+1` sais would suffice to ensure there are at least `Kcp=0` unstable slots before the end of the schema (actually: that there is at least one alternative asai).

Suppose `Kcp=1`.
(Again, the honest schema must have two active slots in the same window of `Scg+1` slots, one of which may be the genesis block.)
This is similar to the `1<Kcp` case, except there might only be one honest active in the first `Scg` sais, and so the first alternative active sai might need to be after the `Scg`th sai, meaning the unstable slots couldn't begin before the `2Scg+2`nd sai.
Also, the `Kcp+1=2`nd honest asai might be as late as the `2Scg`th sai, so the unstable slots might not begin until the `2Scg+Del+1`st sai.
Thus `2Scg + max (Del+2) 3` sais would suffice to ensure there are at least `Kcp=1` unstable slot before the end of the schema.

Since the genesis block can always be chosen as the intersection (as long as the first honest active slot is at most the `Scg`th slot), the above sai counts are also the minimum bound for `Len` in order to ensure at least `Kcp+1` asais for both honest and alternative.